ulib/experimental/Steel.Reference.fsti

```diff
@@ -198,20 +222,61 @@ val free (#a:Type0) (r:ref a) : Steel unit
   (ensures fun _ _ _ -> True)
 
 /// Reads the current value of reference [r]
-val read (#a:Type0) (r:ref a) : Steel a
+val readp (#a:Type0) (r:ref a) (p: perm) : Steel a
+  (vptrp r p) (fun _ -> vptrp r p)
+  (requires fun _ -> True)
+  (ensures fun h0 x h1 -> h0 (vptrp r p) == h1 (vptrp r p) /\ x == h1 (vptrp r p))
+
+let read (#a:Type0) (r:ref a) : Steel a
   (vptr r) (fun _ -> vptr r)
   (requires fun _ -> True)
   (ensures fun h0 x h1 -> sel r h0 == sel r h1 /\ x == sel r h1)
+= readp r full_perm
 
 /// Writes value [x] in reference [r]
 val write (#a:Type0) (r:ref a) (x:a) : Steel unit
   (vptr r) (fun _ -> vptr r)
   (requires fun _ -> True)
   (ensures fun _ _ h1 -> x == sel r h1)
 
+val share (#a:Type0) (#uses:_) (r:ref a) (p: perm)
+  : SteelGhost perm uses
```

Comment:
Is returning the perm needed here? Can we instead return unit, and inline `res == half_perm` in the vprop postcondition?

Reply:
Thank you Aymeric for your feedback so far!

I am proposing to avoid the user remembering the actual structure of the permission values in terms of `half_perm` and `sum_perm`, except maybe for `full_perm`. Consider the following permission associativity example, in the style I am proposing:
```
let dummy () : SteelT unit emp (fun _ -> emp) =
  let r = alloc false in
  let p1 = share r full_perm in
  let p2 = share r p1 in
  let p3 = gather r p1 p2 in
  let p4 = gather r p3 p2 in
  change_equal_slprop (vptrp r p4) (vptrp r full_perm);
  free r
```
However, with your proposal, I believe such a user would need to write:
```
  share r full_perm;
  share r (half_perm full_perm);
  gather r (half_perm full_perm) (half_perm (half_perm full_perm));
  gather r (sum_perm (half_perm full_perm) (half_perm (half_perm full_perm))) (half_perm (half_perm full_perm));
  change_equal_slprop (vptrp r .........) (vptrp r full_perm);
  free r
```

Reply:
I guess I'm mostly wondering about two things:
- How does returning a permission interact with extraction?
- How often do we need to specify the permission when calling share/gather? Why is it not an implicit inferred by the framing tactic, and occasionally rewritten by SMT.

If you do have two consecutive `share` calls, automatic inference will indeed fail. But if you only split the permission in half (which seems to be the most common usecase), no user annotation should be required.

Reply:
> * How does returning a permission interact with extraction?

The `perm` type is an erasable type (by virtue of it being annotated with the `erasable`  attribute, cf. https://github.com/FStarLang/FStar/blob/14b5a29b5631635150cfa07baeb6c9ed6407b767/ulib/experimental/Steel.FractionalPermission.fst#L27-L29), so it will not appear at extraction at all.

> * How often do we need to specify the permission when calling share/gather? Why is it not an implicit inferred by the framing tactic, and occasionally rewritten by SMT.

It seems that `gather` will always need at least one permission to be explicitly given by the user. Here is the minimal amount of annotations in my example:
```
let dummy () : SteelT unit emp (λ _ → emp) =
  let r = alloc false in
  let p1 = share r _ in
  let p2 = share r _ in
  let p3 = gather r p2 _ in
  let p4 = gather r p3 _ in
  change_equal_slprop (vptrp r _) (vptrp r full_perm);
  free r
```

> If you do have two consecutive `share` calls, automatic inference will indeed fail. But if you only split the permission in half (which seems to be the most common usecase), no user annotation should be required.

It alas appears not, as the following example:
```
let dummy1 () : SteelT unit emp (λ _ → emp) =
  let r = alloc false in
  let p1 = share r _ in
  let p3 = gather r _ _ in
  change_equal_slprop (vptrp r _) (vptrp r full_perm);
  free r
```
fails with:
```
(Error) user tactic failed: `could not find a solution for unifying
Steel.Reference.vptrp r x * Steel.Reference.vptrp r x
and
Steel.Reference.vptrp r ((*?u4231*) _ x) * Steel.Reference.vptrp r ((*?u4230*) _ x) * (*?u4229*) _ x`
```

Following your proposal, even if `gather` is constrained to take two halves of the same permission, the following example:
```
assume val share2 (#a:Type0) (#uses:_) (r:ref a) (p: perm)
  : SteelGhostT unit uses (vptrp r p) (λ _ → vptrp r (half_perm p) `star` vptrp r (half_perm p))

assume val gather2 (#a:Type0) (#uses:_)  (r:ref a) (p: perm)
  : SteelGhostT unit uses (vptrp r (half_perm p) `star` vptrp r (half_perm p)) (λ _ → vptrp r p)

let dummy2 () : SteelT unit emp (λ _ → emp) =
  let r = alloc false in
  share2 r _;
  gather2 r _;
  free r
```
fails with:
```
(Error) user tactic failed: `could not find a solution for unifying
Steel.Reference.vptrp r (Steel.FractionalPermission.half_perm Steel.FractionalPermission.full_perm) * Steel.Reference.vptrp r (Steel.FractionalPermission.half_perm Steel.FractionalPermission.full_perm)
and
Steel.Reference.vptrp r (Steel.FractionalPermission.half_perm (*?u8713*) _) * Steel.Reference.vptrp r (Steel.FractionalPermission.half_perm (*?u8713*) _) * (*?u8712*) _`
```

Reply:
There used to be a special case in the framing tactic to handle the cases where all candidates on the left-hand side are identical. This allowed for instance `pts_to r p * pts_to r p` to be unified automatically with `pts_to r ?u1 * pts_to r ?u2`, which corresponds to the share/gather case here.
I'm not sure when this support disappeared, I might have accidentally scrapped it during a merge. I'll investigate and report back,  which should simplify share/gather reasoning here.
My hope is that a user almost never needs to specify permissions when calling share/gather.

Reply:
If I manage to get the Steel automation in this state, would your opinion on returning permissions change?
My impression is that, if we rarely need to use permissions values, having them in the context would be superfluous, but I might be overlooking something.

Reply:
With what you are proposing, I guess calls to `gather` will always automatically mirror calls to `share`. So, we could even replace real number-based fractional permissions with trees, and reason strictly structurally, with no associativity. Is that what we ultimately want, or am I missing anything?

Reply:
Thank you for your patch on bidirectional framing (ad841c6477244024b02236264e551514497de77d), which seems to work very well! Based on that, I changed the interfaces so that:
* `share` no longer returns a permission
* I renamed `gather` into `gather_gen`
* I introduced a more restricted `gather` now operating on half-permissions

See for instance these tests: https://github.com/FStarLang/FStar/blob/a691b6a8ddcd5cc1c0a308887ee7c9f28a9f4013/examples/steel/Selectors.Examples.fst#L63-L78